mmseg/models/backbones/hrformer.py

```diff
@@ -0,0 +1,1071 @@
+        out_dim = self.vdim
+        num_heads = self.num_heads
+        v_head_dim = self.vdim // num_heads
+        assert (head_dim * self.num_heads == embed_dim
+                ), 'embed_dim must be divisible by num_heads'
+        scaling = float(head_dim)**-0.5
+
+        # whether or not use the original query/key/value
+        q = self.q_proj(query) * scaling
+        k = self.k_proj(key)
+        v = self.v_proj(value)
+
+        # divide heads
+        q = q.contiguous().view(tgt_len, bsz * num_heads,
+                                head_dim).transpose(0, 1)
+        if k is not None:
+            k = k.contiguous().view(-1, bsz * num_heads,
+                                    head_dim).transpose(0, 1)
+        if v is not None:
+            v = v.contiguous().view(-1, bsz * num_heads,
+                                    v_head_dim).transpose(0, 1)
+        src_len = k.size(1)
+
+        attn_output_weights = torch.bmm(q, k.transpose(1, 2))
+        assert list(
+            attn_output_weights.size()) == [bsz * num_heads, tgt_len, src_len]
+        """
+        Add relative position embedding
+        """
+        # NOTE: we assume that the src_len == tgt_len == window_size**2
+        assert (
+            src_len == self.window_size[0] * self.window_size[1]
+            and tgt_len == self.window_size[0] * self.window_size[1]
+        ), f'src_len={src_len}, tgt_len={tgt_len},' \
+            f' window_size={self.window_size}'
+        relative_position_bias = self.relative_position_bias_table[
+            self.relative_position_index.view(-1)].view(
+                self.window_size[0] * self.window_size[1],
+                self.window_size[0] * self.window_size[1],
+                -1,
+            )  # Wh*Ww,Wh*Ww,nH
+        relative_position_bias = relative_position_bias.permute(
+            2, 0, 1).contiguous()  # nH, Wh*Ww, Wh*Ww
+        attn_output_weights = attn_output_weights.view(
+            bsz, num_heads, tgt_len,
+            src_len) + relative_position_bias.unsqueeze(0)
+        attn_output_weights = attn_output_weights.view(bsz * num_heads,
+                                                       tgt_len, src_len)
+        """
+        Reweight the attention map before softmax().
+        attn_output_weights: (b*n_head, n, hw)
+        """
+        attn_output_weights = softmax(attn_output_weights, dim=-1)
+        attn_output_weights = dropout(
+            attn_output_weights, p=dropout_p, training=training)
+
+        attn_output = torch.bmm(attn_output_weights, v)
+        assert list(
+            attn_output.size()) == [bsz * num_heads, tgt_len, v_head_dim]
+        attn_output = (
+            attn_output.transpose(0,
+                                  1).contiguous().view(tgt_len, bsz, out_dim))
+        attn_output = linear(attn_output, self.out_proj.weight,
+                             self.out_proj.bias)
+        return attn_output
+
+
+class LocalWindowSelfAttention(nn.Module):
```

Comment:
Can we reuse `LocallyGroupedSelfAttention` in [`twins`](https://github.com/open-mmlab/mmsegmentation/blob/master/mmseg/models/backbones/twins.py#L147) or `WindowMSA` in [`swin`](https://github.com/open-mmlab/mmsegmentation/blob/master/mmseg/models/backbones/swin.py#L22)

Reply:
The same as `GlobalSubsampledAttention` in `twins` reuses EfficientMultiheadAttention in `mit`.